dynatrace/dynatrace.go

```diff
@@ -95,72 +97,154 @@ func defaultFormatter(namespace, name string) string {
+	if kind == aggregation.HistogramKind {
+		return aggregation.DeltaTemporality
+	}
+
+	if desc.InstrumentKind().Adding() && desc.InstrumentKind().Monotonic() {
+		return aggregation.DeltaTemporality
+	}
+
+	return aggregation.CumulativeTemporality
 }
 
 // Export given CheckpointSet
-func (e *Exporter) Export(ctx context.Context, cs export.CheckpointSet) error {
-	// TODO tags
-	// TODO normalize
+func (e *Exporter) Export(ctx context.Context, res *resource.Resource, reader export.InstrumentationLibraryReader) error {
 	lines := []string{}
 
-	err := cs.ForEach(e, func(r export.Record) error {
-		itr := label.NewMergeIterator(r.Labels(), r.Resource().LabelSet())
-		dims := []dimensions.Dimension{}
-		for itr.Next() {
-			label := itr.Label()
-			dims = append(dims, dimensions.NewDimension(string(label.Key), label.Value.Emit()))
-		}
-
-		valOpt, err := getValueOption(r)
-
-		if err != nil {
-			e.logger.Warn(fmt.Sprintf("failed to normalize metric: %s - %s", r.Descriptor().Name(), err.Error()))
-			return nil
-		}
-
-		metric, err := dtMetric.NewMetric(
-			r.Descriptor().Name(),
-			dtMetric.WithPrefix(e.opts.Prefix),
-			dtMetric.WithDimensions(
-				dimensions.MergeLists(
-					e.defaultDimensions,
-					dimensions.NewNormalizedDimensionList(dims...),
-					e.staticDimensions,
-				),
-			),
-			valOpt,
-		)
-
-		if err != nil {
-			e.logger.Warn(fmt.Sprintf("failed to normalize metric: %s - %s", r.Descriptor().Name(), err.Error()))
-			return nil
-		}
+	_ = reader.ForEach(func(l instrumentation.Library, reader export.Reader) error {
+		return reader.ForEach(e, func(record export.Record) error {
+			dims := []dimensions.Dimension{}
+			iter := attribute.NewMergeIterator(record.Labels(), res.Set())
 
-		line, err := metric.Serialize()
+			for iter.Next() {
+				label := iter.Label()
+				dims = append(dims, NewDimension(string(label.Key), label.Value.AsString()))
+			}
 
-		if err != nil {
-			e.logger.Warn(fmt.Sprintf("failed to serialize metric: %s - %s", r.Descriptor().Name(), err.Error()))
+			agg := record.Aggregation()
+
+			dtDimensions := dimensions.MergeLists(
+				e.defaultDimensions,
+				dimensions.NewNormalizedDimensionList(dims...),
+				e.staticDimensions,
+			)
+
+			if hist, ok := agg.(aggregation.Histogram); ok {
+				summary, err := summaryFromHistogram(hist, record.Descriptor().NumberKind())
+
+				if err != nil {
+					e.logger.Sugar().Errorw("error converting histogram to dt summary",
+						"name", record.Descriptor().Name(),
+						"error", err)
+					return nil
+				}
+
+				if summary != nil {
+					metric, err := dtMetric.NewMetric(
+						record.Descriptor().Name(),
+						dtMetric.WithPrefix(e.opts.Prefix),
+						dtMetric.WithDimensions(dtDimensions),
+						summary,
+					)
+
+					if err != nil {
+						e.logger.Sugar().Errorw("error creating summary metric from histogram summary",
+							"name", record.Descriptor().Name(),
+							"error", err)
+					}
```

Comment:
The idea was to be able to return an error from NewMetric that would be logged, but if a metric was also returned it would be exported anyway. Probably best not to deal with such ambiguity though.